README.md
[![Build Status](https://travis-ci.org/magx2/rf24-java-jni.svg?branch=master)](https://travis-ci.org/magx2/rf24-java-jni)

# Installing into your java project

## Get 

### Maven

Add to ```pom.xml```:

```xml
<dependency>
    <groupId>pl.grzeslowski.smarthome</groupId> 
    <artifactId>rf24</artifactId> 
    <version>1.0.0</version> 
    <type>pom</type> 
</dependency>
```

### Gradle

Add to ```build.gradle```:

```gradle
repositories {
	maven {url 'https://dl.bintray.com/big-boy/bigboy'}
}

dependencies {
	compile 'pl.grzeslowski.smarthome:rf24:1.0.0'
}
```

### Jars

You can download Jars from here: 
https://bintray.com/big-boy/bigboy/pl.grzeslowski.smarthome.rf24/1.0.0#files/pl/grzeslowski/smarthome/rf24

## Usage

Create instance of interface ```BasicRf24```

```java
Pins pins = new Pins(ce, cs, clockSpeed);
Retry retry = new Retry(retryDelay, retryNumber);
Payload payload = new Payload((short) (Long.SIZE / Byte.SIZE));
        
BasicRf24 rf24 = new Rf24Adapter(pins, retry, payload);
```

Then ```init``` instance:

```java
rf24.init()
```

And voila! Now you can ```write``` and ```read``` from ```Pipe```s.

```java
rf24.read(READ_PIPE, readBuffer);
rf24.write(WRITE_PIPE, sendBuffer.array());
```

Note: *Remember to ```close``` ```BasicRf24``` instance after you are done.*

For more information look into examples.

# Running tests

Warning: *Remember you need to have ```sudo``` right to run examples!*

## Ping pong (checking round trip time)

In this 2 examples you can run code on 2 different devices (i.e. Java + Java or Java + Arduino) and check what is the round trip time.

### Server

To run it type (in main dir):

    ./gradlew :examples:runRf24PingPongServerExample
    
Note: *You can run Rf24PingPongClientExample.ino on Arduino to have second device that will answer calls.*

### Client

To run it type (in main dir):

    ./gradlew :examples:runRf24PingPongClientExample

Note: *You can run Rf24PingPongServerExample.ino on Arduino to have second device that will start pinging this example.*

# RF24 

## Prerequisites

Note: *You need those programs on machine that you want to compile 
library, i.e. Raspberry PI*

```bash
$ git --version
git version 2.7.4

$ gcc --version
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.2) 5.4.0 20160609
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ g++ --version
g++ (Ubuntu 5.4.0-6ubuntu1~16.04.2) 5.4.0 20160609
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Install everything with:

```bash
sudo apt-get install git gcc g++
```

Also remember to update your system:

```bash
sudo apt-get update
sudo apt-get dist-upgrade
```

### Enable SPI on Raspberry PI

Open RPi config app ```sudo raspi-config```, select Advanced and enable 
the SPI kernel module. ```sudo reboot```

### Clone repository

```bash
git clone https://github.com/magx2/rf24-java-jni.git
cd rf24-java-jni
```

## RF24 native libs

Native libraries is taken from this repository: https://github.com/TMRh20/RF24/
(checked with commit SHA **c94f87746552a3c662c35fd11e772b747f19d8ee**). 
This repository is in folder *./native/RF24* as a ```git submodule```. To download
it run:

```bash
git submodule init
git submodule update
ls ./native/RF24
```

## Compiling native library

Note: *In ./native/precompiled/ folder you can fins precomputed libraries - 
there is no need to compile it by hand!*

On you RPi just run (remember you need to have ```sudo``` rights!):

```gradle
./gradlew clean compileRf24
```

If everything went smooth you should find your native library in *./build/wrapper/librf24bcmjava.so*.
Use this library with your app. Remember to load it in your java code before using any RF24 related code:

```java
public class Main {
    static {
        try {
            System.loadLibrary("rf24bcmjava");
        } catch (UnsatisfiedLinkError e) {
            // handel this if you want
        }
    }

    public static void main(String[] args) {
        // ... your staff ...
    }
}
```

If you will be using my ```RF24Adapter``` use provided method to do this 
less verbose: ```RF24Adapter.loadLibrary()```.

### Configuring your build

There are some properties that you can configure. All defaults runs perfectly fine
on Raspberry PI 3 with Raspbian.

Variable                | Description                     | Default
------------------------|---------------------------------|---------
javaJni                 | Points where headers fo JNI are | /usr/lib/jvm/jdk-8-oracle-arm32-vfp-hflt/include
javaJniPlatformSpecific | Points where platform specific headers fo JNI are. | $javaJni/linux
headersDir              | Remote header files target installation directory. Run ```./native/RF24/configure --help``` for more information | RPi
wrapperBuildDir         | Where to build wrapper           | $project.buildDir/wrapper

To set those properties you can run with ```-P<var>=<new_value>``` or
 set it in *~/.gradle/gradle.properties* file.

# SWIG

Note: *This step is not mandatory. All artefacts of SWIG compilation are commited
into repository (./swig-config/rf24bcmjava_wrap.cxx and generated classes in
./src/main/java/pl/grzeslowski/smarthome/rf24/generated).*

## Install SWIG

Download SWIG from http://www.swig.org/download.html (tested on version 3.0.10)
and install it. In downloaded zip you should find file called *INSTALL*. 
Follow this instructions.

Note: *Remember to add swig folder into your PATH (edit ~/.bashrc file on 
Linux)*

Check if your installation has succed:

```bash
$ swig -version
SWIG Version 3.0.10
Compiled with g++ [x86_64-pc-linux-gnu]
Configured options: +pcre
```

## Generate SWIG files

First run Gradle task to clean SWIG, ```removeSwigGeneratedFiles``` and then
run code generation ```compileSwig``` 

```gradle
./gardlew removeSwigGeneratedFiles compileSwig
```

Note: *If you want to commit those newly generated files, just fork my repository.
Also think if it would be good to make pull request back to me :).*

# Tested on

## Compiling RF24

Raspberry PI 3 with Raspbian

```bash
$ uname -a
Linux raspberrypi 4.4.21-v7+ #911 SMP Thu Sep 15 14:22:38 BST 2016 armv7l GNU/Linux
```

## Compiling SWIG

PC with Ubuntu:

```bash
$ uname -a
Linux xxx 4.4.0-45-generic #66-Ubuntu SMP Wed Oct 19 14:12:37 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```


RELEASE_NOTES.md
# Release notes

## In progress...
 
* Added method ```isInit()``` to check whenever object was already initialized

## 16-11-18 v1.0.0

* Initial version
* Implemented BasicRf24 
* Added ping pong examples

src/main/java/pl/grzeslowski/smarthome/rf24/rpi/RpiGpio.java
package pl.grzeslowski.smarthome.rf24.rpi;

import java.util.Optional;

import static java.util.Arrays.stream;

/**
 * Convenient way to use RPi GPIO pins with RF24 devices.
 *
 * {@see pl.grzeslowski.smarthome.rf24.rpi.RpiPins}
 */
public enum RpiGpio {
    /* Version 1, Pin P1-03 */

    RPI_GPIO_P1_03((short) 0),
    /**
     * Version 1, Pin P1-05
     */
    RPI_GPIO_P1_05((short) 1),
    /**
     * Version 1, Pin P1-07
     */
    RPI_GPIO_P1_07((short) 4),
    /**
     * Version 1, Pin P1-08, defaults to alt function 0 UART0_TXD
     */
    RPI_GPIO_P1_08((short) 14),
    /**
     * Version 1, Pin P1-10, defaults to alt function 0 UART0_RXD
     */
    RPI_GPIO_P1_10((short) 15),
    /**
     * Version 1, Pin P1-11
     */
    RPI_GPIO_P1_11((short) 17),
    /**
     * Version 1, Pin P1-12, can be PWM channel 0 in ALT FUN 5
     */
    RPI_GPIO_P1_12((short) 18),
    /**
     * Version 1, Pin P1-13
     */
    RPI_GPIO_P1_13((short) 21),
    /**
     * Version 1, Pin P1-15
     */
    RPI_GPIO_P1_15((short) 22),
    /**
     * Version 1, Pin P1-16
     */
    RPI_GPIO_P1_16((short) 23),
    /**
     * Version 1, Pin P1-18
     */
    RPI_GPIO_P1_18((short) 24),
    /**
     * Version 1, Pin P1-19, MOSI when SPI0 in use
     */
    RPI_GPIO_P1_19((short) 10),
    /**
     * Version 1, Pin P1-21, MISO when SPI0 in use
     */
    RPI_GPIO_P1_21((short) 9),
    /**
     * Version 1, Pin P1-22
     */
    RPI_GPIO_P1_22((short) 25),
    /**
     * Version 1, Pin P1-23, CLK when SPI0 in use
     */
    RPI_GPIO_P1_23((short) 11),
    /**
     * Version 1, Pin P1-24, CE0 when SPI0 in use
     */
    RPI_GPIO_P1_24((short) 8),
    /**
     * Version 1, Pin P1-26, CE1 when SPI0 in use
     */
    RPI_GPIO_P1_26((short) 7),


    /* RPi Version 2 */

    /**
     * Version 2, Pin P1-03
     */
    RPI_V2_GPIO_P1_03((short) 2),
    /**
     * Version 2, Pin P1-05
     */
    RPI_V2_GPIO_P1_05((short) 3),
    /**
     * Version 2, Pin P1-07
     */
    RPI_V2_GPIO_P1_07((short) 4),
    /**
     * Version 2, Pin P1-08, defaults to alt function 0 UART0_TXD
     */
    RPI_V2_GPIO_P1_08((short) 14),
    /**
     * Version 2, Pin P1-10, defaults to alt function 0 UART0_RXD
     */
    RPI_V2_GPIO_P1_10((short) 15),
    /**
     * Version 2, Pin P1-11
     */
    RPI_V2_GPIO_P1_11((short) 17),
    /**
     * Version 2, Pin P1-12, can be PWM channel 0 in ALT FUN 5
     */
    RPI_V2_GPIO_P1_12((short) 18),
    /**
     * Version 2, Pin P1-13
     */
    RPI_V2_GPIO_P1_13((short) 27),

    /**
     * Version 2, Pin P1-15
     */
    RPI_V2_GPIO_P1_15((short) 22),
    /**
     * Version 2, Pin P1-16
     */
    RPI_V2_GPIO_P1_16((short) 23),
    /**
     * Version 2, Pin P1-18
     */
    RPI_V2_GPIO_P1_18((short) 24),
    /**
     * Version 2, Pin P1-19, MOSI when SPI0 in use
     */
    RPI_V2_GPIO_P1_19((short) 10),
    /**
     * Version 2, Pin P1-21, MISO when SPI0 in use
     */
    RPI_V2_GPIO_P1_21((short) 9),
    /**
     * Version 2, Pin P1-22
     */
    RPI_V2_GPIO_P1_22((short) 25),
    /**
     * Version 2, Pin P1-23, CLK when SPI0 in use
     */
    RPI_V2_GPIO_P1_23((short) 11),
    /**
     * Version 2, Pin P1-24, CE0 when SPI0 in use
     */
    RPI_V2_GPIO_P1_24((short) 8),
    /**
     * Version 2, Pin P1-26, CE1 when SPI0 in use
     */
    RPI_V2_GPIO_P1_26((short) 7),
    /**
     * Version 2, Pin P1-29
     */
    RPI_V2_GPIO_P1_29((short) 5),
    /**
     * Version 2, Pin P1-31
     */
    RPI_V2_GPIO_P1_31((short) 6),
    /**
     * Version 2, Pin P1-32
     */
    RPI_V2_GPIO_P1_32((short) 12),
    /**
     * Version 2, Pin P1-33
     */
    RPI_V2_GPIO_P1_33((short) 13),
    /**
     * Version 2, Pin P1-35
     */
    RPI_V2_GPIO_P1_35((short) 19),
    /**
     * Version 2, Pin P1-36
     */
    RPI_V2_GPIO_P1_36((short) 16),
    /**
     * Version 2, Pin P1-37
     */
    RPI_V2_GPIO_P1_37((short) 26),
    /**
     * Version 2, Pin P1-38
     */
    RPI_V2_GPIO_P1_38((short) 20),
    /**
     * Version 2, Pin P1-40
     */
    RPI_V2_GPIO_P1_40((short) 21),


    /* RPi Version 2, new plug P5 */

    /**
     * Version 2, Pin P5-03
     */
    RPI_V2_GPIO_P5_03((short) 28),
    /**
     * Version 2, Pin P5-04
     */
    RPI_V2_GPIO_P5_04((short) 29),
    /**
     * Version 2, Pin P5-05
     */
    RPI_V2_GPIO_P5_05((short) 30),
    /**
     * Version 2, Pin P5-06
     */
    RPI_V2_GPIO_P5_06((short) 31),


    /* RPi B+ J8 header, also RPi 2 40 pin GPIO header */

    /**
     * B+, Pin J8-03
     */
    RPI_BPLUS_GPIO_J8_03((short) 2),
    /**
     * B+, Pin J8-05
     */
    RPI_BPLUS_GPIO_J8_05((short) 3),
    /**
     * B+, Pin J8-07
     */
    RPI_BPLUS_GPIO_J8_07((short) 4),
    /**
     * B+, Pin J8-08, defaults to alt function 0 UART0_TXD
     */
    RPI_BPLUS_GPIO_J8_08((short) 14),
    /**
     * B+, Pin J8-10, defaults to alt function 0 UART0_RXD
     */
    RPI_BPLUS_GPIO_J8_10((short) 15),
    /**
     * B+, Pin J8-11
     */
    RPI_BPLUS_GPIO_J8_11((short) 17),
    /**
     * B+, Pin J8-12, can be PWM channel 0 in ALT FUN 5
     */
    RPI_BPLUS_GPIO_J8_12((short) 18),
    /**
     * B+, Pin J8-13
     */
    RPI_BPLUS_GPIO_J8_13((short) 27),
    /**
     * B+, Pin J8-15
     */
    RPI_BPLUS_GPIO_J8_15((short) 22),
    /**
     * B+, Pin J8-16
     */
    RPI_BPLUS_GPIO_J8_16((short) 23),
    /**
     * B+, Pin J8-18
     */
    RPI_BPLUS_GPIO_J8_18((short) 24),
    /**
     * B+, Pin J8-19, MOSI when SPI0 in use
     */
    RPI_BPLUS_GPIO_J8_19((short) 10),
    /**
     * B+, Pin J8-21, MISO when SPI0 in use
     */
    RPI_BPLUS_GPIO_J8_21((short) 9),
    /**
     * B+, Pin J8-22
     */
    RPI_BPLUS_GPIO_J8_22((short) 25),
    /**
     * B+, Pin J8-23, CLK when SPI0 in use
     */
    RPI_BPLUS_GPIO_J8_23((short) 11),
    /**
     * B+, Pin J8-24, CE0 when SPI0 in use
     */
    RPI_BPLUS_GPIO_J8_24((short) 8),
    /**
     * B+, Pin J8-26, CE1 when SPI0 in use
     */
    RPI_BPLUS_GPIO_J8_26((short) 7),
    /**
     * B+, Pin J8-29,
     */
    RPI_BPLUS_GPIO_J8_29((short) 5),
    /**
     * B+, Pin J8-31,
     */
    RPI_BPLUS_GPIO_J8_31((short) 6),
    /**
     * B+, Pin J8-32,
     */
    RPI_BPLUS_GPIO_J8_32((short) 12),
    /**
     * B+, Pin J8-33,
     */
    RPI_BPLUS_GPIO_J8_33((short) 13),
    /**
     * B+, Pin J8-35,
     */
    RPI_BPLUS_GPIO_J8_35((short) 19),

    /**
     * B+, Pin J8-36,
     */
    RPI_BPLUS_GPIO_J8_36((short) 16),
    /**
     * B+, Pin J8-37,
     */
    RPI_BPLUS_GPIO_J8_37((short) 26),
    /**
     * B+, Pin J8-38,
     */
    RPI_BPLUS_GPIO_J8_38((short) 20),
    /**
     * B+, Pin J8-40,
     */
    RPI_BPLUS_GPIO_J8_40((short) 21);


    private final short gpioPin;

    RpiGpio(short gpioPin) {
        this.gpioPin = gpioPin;
    }

    public short getGpioPin() {
        return gpioPin;
    }

    public static Optional<RpiGpio> fingForGpioPin(short gpioPin) {
        return stream(RpiGpio.values())
                .filter(gpio -> gpio.gpioPin == gpioPin)
                .findFirst();
    }

}
